tests/draft2020-12/dynamicRef.json

```diff
@@ -253,5 +253,109 @@
                 "valid": false
             }
         ]
+    },
+    {
+        "description": "dynamicRefs with relative uris and defs",
+        "schema": {
+            "$id": "schema1.json",
+            "properties": {
+                "foo": {
+                    "$id": "schema2.json",
+                    "$defs": {
+                        "inner": {
+                            "properties": {
+                                "bar": { "type": "string" }
+                            }
+                        }
+                    },
+                    "$dynamicRef": "#/$defs/inner"
```

Comment:
Looking at the draft, it says:

> The value of the "$dynamicRef" property MUST be a string which is a URI-Reference. Resolved against the current URI base, it produces the URI used as the starting point for runtime resolution. This initial resolution is safe to perform on schema load.
> 
> If  the initially resolved starting point URI includes a fragment that was created by the "$dynamicAnchor" keyword, the initial URI MUST be replaced by the URI (including the fragment) for the outermost schema resource in the dynamic scope that defines an identically named fragment with "$dynamicAnchor".
> 
> Otherwise, its behavior is identical to "$ref", and no runtime resolution is needed. 

I didn't see any place which limits the URI to a plain identifier. I can remove this test if you find it better, but I think this is the correct behavior as defined by the current draft.

Reply:
I think you're right. It's nonsensical, but not technically wrong according to the spec. You can't expect any dynamic behavior because `$dynamicRef`s work with `$dynamicAnchor`s which can only be plain identifiers. Because the URI fragment can't match a `$dynamicAnchor`, it behaves like a `$ref` where that syntax is valid. That's terrible and not what the "identical to `$ref`" clause was intended for. You found a bizarre edge case that wasn't considered when the spec was written.

In any case, this test isn't supposed to be testing an edge case, it's supposed to be testing relative URI resolution. So, I'd rather see this test use `$dynamicRef` as it was intended. It will make more sense and it will age better.

Reply:
To elaborate a little more on the "identical to `$ref`" clause, that was meant as a fallback for when a `$dynamicAnchor` was not found that matches the URI fragment. It was not intended to allow `$dynamicRef` to be used in ways that weren't trying to do a dynamic reference.

`$dynamicRef` has some baggage it inherited from `$recursiveRef` that I think is the problem. Not everything that made sense for `$recursiveRef` makes sense for `$dynamicRef`. I'm trying to get it fixed for the next draft.

Reply:
I think the best tests are for strange edge cases :) But if you think this an unimportant edge case which will get removed in the next draft, I can remove this; in that case, that draft should probably explicitly say that `$dynamicRef` is undefined when a matching `$dynamicAnchor` is not found, that it should only be a fragment, etc. (like `$recursiveRef` could only be the empty fragment, i.e. `#`)

As for a version which uses `$dynamicRef` "properly", I would suggest something like this:
```json
{
"$id": "schema1.json",
"properties": {
    "foo": {
        "$id": "schema2.json",
        "$dynamicAnchor": "foo"
        "$defs": {
            "inner": {
                "$dynamicAnchor": "inner"
                "properties": {
                    "bar": { "type": "string" }
                }
            }
        },
        "$dynamicRef": "#inner"
    },
    "$dynamicRef": "#foo"
}
```
And a similar one could be written with normal `$anchor`-s. However, at this point I feel this is too far removed from the original idea of this test, so I'm not sure whether it's relevant to include here. What do you think?

Reply:
> I think the best tests are for strange edge cases :) But if you think this an unimportant edge case which will get removed in the next draft, I can remove this

I don't think it needs a test, but that's not the point. The argument isn't that the edge case is not important or shouldn't be tested, it's that a test for an edge case shouldn't be lumped in with this test that has nothing to do with that edge case. A good test should test one thing and have one reason to fail. This test shouldn't fail because of a bug handling this edge case.

The alternate version looks reasonable to me.

Reply:
Ok, as above, I'll make sure that the "split" test finds the relevant bugs, and then update.

Reply:
I checked. My "compound" version above (the original) does indeed capture a bug that the "split" version doesn't. So I would prefer to add that test as-is. I can still add all the other tests written, which include separate ones for various `$dynamicRef`-s without `$dynamicAnchor`-s, and the one which uses them with an anchor.

Reply:
This is taking too long and getting off-topic. I decided to completely remove the `$dynamicRef` tests from this PR, so it can be merged; once it lands, I'll open a new one where we can discuss solely those tests.